dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DefaultDeduplicationService.java

```diff
@@ -106,27 +106,35 @@ public void updatePotentialDuplicate( PotentialDuplicate potentialDuplicate )
 
     @Override
     @Transactional
-    public void autoMerge( DeduplicationMergeRequest deduplicationMergeRequest )
+    public void autoMerge( DeduplicationMergeParams params )
     {
-        if ( !isAutoMergeable( deduplicationMergeRequest.getOriginal(), deduplicationMergeRequest.getDuplicate() ) )
-            throw new PotentialDuplicateConflictException( "Potential Duplicate is not automatically mergeable" );
+        if ( !isAutoMergeable( params.getOriginal(), params.getDuplicate() ) )
+            throw new PotentialDuplicateConflictException( "PotentialDuplicate can not be merged automatically." );
 
-        MergeObject mergeObject = deduplicationHelper.generateMergeObject( deduplicationMergeRequest.getOriginal(),
-            deduplicationMergeRequest.getDuplicate() );
-        deduplicationMergeRequest.setMergeObject( mergeObject );
-        merge( deduplicationMergeRequest );
+        params.setMergeObject( deduplicationHelper.generateMergeObject( params.getOriginal(), params.getDuplicate() ) );
+
+        merge( params );
     }
 
     @Override
-    public void manualMerge( DeduplicationMergeRequest deduplicationMergeRequest )
+    @Transactional
+    public void manualMerge( DeduplicationMergeParams deduplicationMergeParams )
     {
-        throw new RuntimeException( "Manual merge not yet implemented" );
+        if ( deduplicationHelper.hasInvalidReferences( deduplicationMergeParams ) )
+        {
+            throw new PotentialDuplicateConflictException(
+                "The submitted payload contains invalid references and cannot be merged." );
+        }
```

Comment:
I know there is never time for this, but we now need to add tests for our business logic.
In DeduplicationServiceTest, when we hit manualMerge we need to check hasInvalidReferences has done its job with something like :

        verify( deduplicationHelper ).hasInvalidReferences( deduplicationMergeParams );

Also, we need to add a new test to check the exception has been caught, something like :  

....
        when( deduplicationHelper.manualMerge( deduplicationMergeParams ) )
            .thenReturn( false );

        assertThrows(
            PotentialDuplicateConflictException.class,
            () -> deduplicationService.manualMerge( deduplicationMergeParams ) );
....

This will help other people working on same code and it is the only way we know our code works indipendently from the actual implementation

Reply:
Thanks for the feedback Luca! I'll be adding the test today :)